### PR TITLE
Reset QuicheMemSliceImpl() in move assignment operator.

### PR DESCRIPTION
**Pull request by bencebeky**
Signed-off-by: Bence Béky <bnc@google.com>

Commit Message: Reset QuicheMemSliceImpl() in move assignment operator.
Additional Description: minor tweak
Risk Level: low
Testing: n/a
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/quic/platform/quiche_mem_slice_impl.h [hunk 1]
```diff
@@ -42,6 +42,7 @@ class QuicheMemSliceImpl {
   QuicheMemSliceImpl& operator=(const QuicheMemSliceImpl& other) = delete;
   QuicheMemSliceImpl& operator=(QuicheMemSliceImpl&& other) noexcept {
     if (this != &other) {
+      Reset();
```
Comment (RyanTheOptimist) [inline source/common/quic/platform/quiche_mem_slice_impl.h:45]: While you're here, would you mind adding a comment about the move on line 47 behaves differently from the move one line 46, which necessitates the Reset() here?

**Review by RyanTheOptimist — COMMENTED** [review]
This LGTM, but as long as you're here it would be awesome to include some of the digging you did into this bug into a comment. Also would it be possible to add a regression test for this bug to envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc?

File: source/common/quic/platform/quiche_mem_slice_impl.h (hunk 1 above)
Comment (bencebeky) [reply source/common/quic/platform/quiche_mem_slice_impl.h:48]: Done.

**Comment by bencebeky** [comment]
> This LGTM, but as long as you're here it would be awesome to include some of the digging you did into this bug into a comment. Also would it be possible to add a regression test for this bug to envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc?

Thank you for the review.

I'm planning to add a test to quiche/common/platform/api/quiche_mem_slice_test.cc in upstream QUICHE.  I couldn't do it before fixing the bug because it would otherwise block rolling QUICHE into Envoy.  This behavior is not implementation-specific, and every QUICHE embedder deserves a test.  In my mind envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc should only be used for tests that are specific to Envoy's platform implementations (or use test machinery that is specific to Envoy), and not for testing platform API in general.

**Comment by RyanTheOptimist** [comment]
> > This LGTM, but as long as you're here it would be awesome to include some of the digging you did into this bug into a comment. Also would it be possible to add a regression test for this bug to envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc?
> 
> Thank you for the review.
> 
> I'm planning to add a test to quiche/common/platform/api/quiche_mem_slice_test.cc in upstream QUICHE. I couldn't do it before fixing the bug because it would otherwise block rolling QUICHE into Envoy. This behavior is not implementation-specific, and every QUICHE embedder deserves a test. In my mind envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc should only be used for tests that are specific to Envoy's platform implementations (or use test machinery that is specific to Envoy), and not for testing platform API in general.

Sounds great. Thanks!

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by bencebeky** [comment]
> > > This LGTM, but as long as you're here it would be awesome to include some of the digging you did into this bug into a comment. Also would it be possible to add a regression test for this bug to envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc?
> > 
> > 
> > Thank you for the review.
> > I'm planning to add a test to quiche/common/platform/api/quiche_mem_slice_test.cc in upstream QUICHE. I couldn't do it before fixing the bug because it would otherwise block rolling QUICHE into Envoy. This behavior is not implementation-specific, and every QUICHE embedder deserves a test. In my mind envoyproxy/envoy/test/common/quic/platform/quic_platform_test.cc should only be used for tests that are specific to Envoy's platform implementations (or use test machinery that is specific to Envoy), and not for testing platform API in general.
> 
> Sounds great. Thanks!

I just added the tests to QUICHE at https://quiche.googlesource.com/quiche/+/e5c727cc55dcd855a44034424c84b4a4a7d1135a.  Note that Envoy build files have a minor bug causing these tests not to run, so this does not block QUICHE roll.  I'm fixing that bug at https://github.com/envoyproxy/envoy/pull/21013.

**Pull request merged**